### PR TITLE
Add Unit Tests to `legacy/src/service.rs`

### PR DESCRIPTION
**Pull request by Ayiga**
### This PR:

Adds unit tests to the various functions and methods contained with the `service.rs` file of the `legacy` crate, formely known as the `hotshot-builder-core` crate.  These tests serve as test cases that assert current behavior, and attempt to draw attention to potential edge case behavior that can lead to confusion in the future.

These unit tests just assert that the currently coded behavior is the expected behavior, without knowing whether or not this is actually the intended case or not.

### This PR does not:

Make any assumptions as whether the current implemented behavior is correct, or if they consider their edge cases correctly.

### Key places to review:

- `test_global_state_register_builder_state_same_builder_state_id`

  This test asserts that submitting the same `BuilderStateId` to `register_builder_state` **SHOULD** result in the replacement of the previously stored Request Sender, and the previous one will be dropped.

- `test_global_state_update_global_state_replacement`

  This test asserts that a subsequent call to `update_global_state` with the same `BuilderStateId` **SHOULD NOT** replace the data currently stored in the `blocks` LRU, and **SHOULD** replace the data stored within `builder_state_to_last_built_block`.

### CODE REVIEW

File: crates/legacy/src/service.rs [hunk 1]
```diff
@@ -1235,15 +1778,3188 @@ pub(crate) async fn handle_received_txns<Types: NodeType>(
...
+    }
+
+    /// This test checks that the error [HandleReceivedTxnsError::Internal]
+    /// is returned when the broadcast channel is closed.
+    ///
+    /// To trigger this error we simply close the broadcast channel receiver
+    /// before attempting to send any transactions through the broadcast channel
+    /// sender.
+    #[async_std::test]
+    async fn test_handle_received_txns_error_something() {
```
Comment (shenkeyao) [inline crates/legacy/src/service.rs:4666]: Nit: Why "something" here?

File: crates/legacy/src/service.rs [hunk 2]
```diff
@@ -1235,15 +1778,3188 @@ pub(crate) async fn handle_received_txns<Types: NodeType>(
...
+        }
+    }
+
+    /// This test checks that the error [ConnectToEventsServiceError::Connection]
+    /// is returned when [connect_to_events_service] is invoked.
+    ///
+    /// This error may be difficult to control in all environments. In order
+    /// to control this specific case, the underlying client being passed to
+    /// [connect_to_events_service] is a mock that always returns false when
+    /// [check_connection] is invoked.
```
Comment (shenkeyao) [inline crates/legacy/src/service.rs:4763]: Nice testing setup. 👍

**Review by shenkeyao — CHANGES_REQUESTED** [review]
Requested some changes but this is really clear and comprehensive.

File: crates/legacy/src/service.rs [hunk 3]
```diff
@@ -234,7 +301,18 @@ impl<Types: NodeType> GlobalState<Types> {
...
+    /// [remove_handles] cleans up the [GlobalState] by removing all
+    /// `spawned_builder_states` that have been stored, up to a derived
+    /// reference view.  This cutoff point can be up to the given
+    /// `on_decide_view` so long as the provided value is less than or equal
+    /// to the `highest_view_num_builder_id`'s view stored on the state.
+    /// Beyond that, the state prefers to drop all `spawned_builder_states`
+    /// preceding the derived cutoff view.
+    ///
+    /// In addition, for tracking purposes, the `last_garbage_collected_view_num`
+    /// is updated to the target cutoff view number for tracking purposes.  The
```
[the remark this answers is not in the document]
Comment (Ayiga) [reply crates/legacy/src/service.rs]: Addressed in 45654a34372259e26d7c53cee01e997832216a92

File: crates/legacy/src/service.rs [hunk 4]
```diff
@@ -1235,15 +1778,3188 @@ pub(crate) async fn handle_received_txns<Types: NodeType>(
...
+            );
+        }
+    }
+
+    // Get Available Blocks Tests
+
+    /// This test checks that the error `AvailableBlocksError::NoBlocksAvailable`
+    /// is returned when no blocks are available.
+    ///
+    /// To Trigger this condition, we simply submit a request to the
```
[the remark this answers is not in the document]
Comment (Ayiga) [reply crates/legacy/src/service.rs]: Addressed in 06211c851aea5a7db16c497376f9ee0aa49a1dbf.

File: crates/legacy/src/service.rs (hunk 1 above)
Comment (Ayiga) [reply crates/legacy/src/service.rs]: Ah, this is a correction I missed.  Essentially an off-by-1 error.  When I start making the specific test to test for an error case, I'm not necessarily certain which error I'm testing for.  As a result, I use a temporary placeholder name.  In this case I failed to change the name so the temporary name became a permanent name.

Addressed in f15ede4c5e99fc012271066da8c4e6fba4ce3e13.

File: crates/legacy/src/service.rs [hunk 5]
```diff
@@ -1235,15 +1778,3188 @@ pub(crate) async fn handle_received_txns<Types: NodeType>(
...
+    }
+
+    /// This test checks that the error `AvailableBlocksError::GetChannelForMatchingBuilderError`
+    /// is returned when attempting to retrieve a view that is not stored within the state, and
+    /// the highest view is also no longer stored within the state.
+    ///
+    /// To trigger this condition, we initialize the GlobalState with an initial
+    /// state, and then we mutate the state to record the wrong latest state id.
+    /// When interacted with `GlobalState` via `register_builder_state`, and
+    /// `remove_handles`, this error doesn't seem possible immediately possible.
```
[the remark this answers is not in the document]
Comment (Ayiga) [reply crates/legacy/src/service.rs:3092]: I'm attempting to state that in order to trigger this condition you must be mutating the internal structure of `GlobalState` directly.  If you instead access it indirectly using `register_builder_state` and the cleanup method `remove_handles` then I don't believe the error is possible to encounter.

In other words.  If you use the `GlobalState` methods as intended instead of the internal structure of the `struct`, this issue does not seem to be able to be triggered.

File: crates/legacy/src/service.rs [hunk 6]
```diff
@@ -1235,15 +1778,3188 @@ pub(crate) async fn handle_received_txns<Types: NodeType>(
...
+    // producible as the unbounded channel doesn't seem to be able to be
+    // closed.
+    //
+    // The second is SigningBlockFailed, which doesn't seem to be producible
+    // with a valid private key, and it's not clear how to create an invalid
+    // private key.
+
+    /// This test checks that call to `available_blocks_implementation` returns
+    /// a successful response when the function is called before blocks are
+    /// made available.
```
[the remark this answers is not in the document]
Comment (Ayiga) [reply crates/legacy/src/service.rs:3160]: I believe this may be true in general.  I wasn't certain at the time I wrote the test.  I do think if this assumption were relaxed and not codified with a test that we could potentially change this behavior so that it does affect this behavior.  However, I don't know how likely that would be.

File: crates/legacy/src/service.rs [hunk 7]
```diff
@@ -1089,16 +1419,56 @@ pub async fn run_permissioned_standalone_builder_service<
...
-*/
+/// [HandleDaEventError] represents the internal class of errors that can
+/// occur when attempting to process an incoming da proposal event.  More
+/// specifically these are the class of error that can be returned from
+/// [handle_da_event_implementation].
+#[derive(Debug)]
+enum HandleDaEventError<Types: NodeType> {
+    SenderIsNotLeader,
+    SignatureValidationFailed,
+    BroadcastFailed(async_broadcast::SendError<MessageType<Types>>),
```
[the remark this answers is not in the document]
Comment (Ayiga) [reply crates/legacy/src/service.rs:1385]: We could probably do this.  Though if we introduce a new error edge case it might occur in one of the functions but not the other.

I suppose the question I have is:  Is this a desired behavior?  Or is it better to have a separate `enum` of `variant` errors per method, even if many or all of the cases are duplicated?

File: crates/legacy/src/service.rs [hunk 8]
```diff
@@ -184,6 +228,18 @@ impl<Types: NodeType> GlobalState<Types> {
...
+    ///
+    /// Additionally, if the view of the [BuilderStateId] is greater than the
+    /// current highest view number, the [BuilderStateId] is set as the new
+    /// highest view number.
+    ///
+    /// There is potential here for data loss.  Since we just blindly insert
+    /// the [BuilderStateId] and [BroadcastSender] into the hashmap, we could
+    /// potentially be overwriting an existing entry.  This would result in
+    /// the loss of access to a [BroadcastSender], and could potentially
+    /// result in unexpected behavior.
```
[the remark this answers is not in the document]
Comment (Ayiga) [reply crates/legacy/src/service.rs:197]: Addressed in 6702d50e9a9ca8da7dab6555c081ead55dc113fb.

File: crates/legacy/src/service.rs [hunk 9]
```diff
@@ -205,6 +261,17 @@ impl<Types: NodeType> GlobalState<Types> {
...
+    /// is within the [GlobalState]'s [blocks] LRU Cache.  The cache stores the
+    /// [BlockInfo] associated with the given [BuildBlockInfo]'s id.  However
+    /// if it already exists within the LRU cache, then the `BlockInfo` is not
+    /// updated.
+    ///
+    /// Additionally, the [BuilderStateId] is associated with the given
+    /// [ResponseMessage] in the [builder_state_to_last_built_block] hashmap.
+    ///
+    /// No care or consideration is given to anything that may have been
+    /// stored with the same key in the [builder_state_to_last_built_block].
```
[the remark this answers is not in the document]
Comment (Ayiga) [reply crates/legacy/src/service.rs:237]: Addressed in e4d7ba846c3351b37dbf078882b1c32597499c09.

File: crates/legacy/src/service.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (shenkeyao) [reply crates/legacy/src/service.rs:1407]: Generally, I'd prefer separate enums, but here the quorum and DA proposal handling are very similar, and even if one has an error case that the other doesn't, we could have a `HandleProposalEventError::QuorumProposalTooLarge` (just a made-up example) variant to make it explicit that it's for the quorum proposal.

However, I don't have a strong opinion. The current way might make debugging easier. I think it's also fine to leave it as is.

File: crates/legacy/src/service.rs [hunk 10]
```diff
@@ -1235,15 +1778,3188 @@ pub(crate) async fn handle_received_txns<Types: NodeType>(
...
+        for i in 0..10 {
+            let builder_state_id = BuilderStateId {
+                parent_commitment: vid_commitment(&[i], 4),
+                view: ViewNumber::new(i as u64),
+            };
+            assert!(
+                !state.spawned_builder_states.contains_key(&builder_state_id),
+                "the spawned builder states should not contain the builder state id, {builder_state_id}"
+            );
+        }
```
[the remark this answers is not in the document]
Comment (shenkeyao) [reply crates/legacy/src/service.rs:2655]: Just checking if this comment was missed. :)

File: crates/legacy/src/service.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (shenkeyao) [reply crates/legacy/src/service.rs:3114]: Makes sense!

File: crates/legacy/src/service.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (shenkeyao) [reply crates/legacy/src/service.rs:3182]: Actually, do we want `available_blocks_implementation` to return the blocks if it's called before blocks are made available? This test passes because we call `await` after the blocks are available, but if we change it to be really "before", won't we expect no available blocks to be returned? I feel it's a bit counter-intuitive to expect `available_blocks_implementation` to give the same result in both cases.

[the remark this answers is not in the document]
Comment (Ayiga) [reply crates/legacy/src/service.rs:3182]: Well, in this case we're dealing with non-synchronous processes.  The `available_blocks_implementation` does have a loop where it attempts to wait for a matching builder with a while loop using this variable: `time_to_wait_for_matching_builder`.  So it definitely seems like there's an expectation that we should be able to receive a request for the block, before producing the block and being able to fulfill this.

Due to the async nature of the API requests versus the receiving of the Proposal events I think this makes sense, as we cannot guarantee that by the time `HotShot` calls the `Builder`'s endpoint that we will have already receive a `Proposal` from the HotShot `Event Stream`.

[the remark this answers is not in the document]
Comment (shenkeyao) [reply crates/legacy/src/service.rs:3182]: Makes sense! I missed the fact that `available_blocks_implementation` was wrapped by `async_spawn(async move {...})` so the task had been started before calling `await`. Agree that the order matters in this async case.

File: crates/legacy/src/service.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (Ayiga) [reply crates/legacy/src/service.rs:2653]: I did miss this one!  Hmm... The previous entry does ensure that there is only one entry stored within the `spawned_builder_states`, however it doesn't test **which** entry remains.  This for-loop is testing to ensure that all of the entries that *should* have been removed are no longer preset within the storage.

We *could* potentially replace this with a check for the `i = 10`th entry only to achieve the same thing.  But I don't think checking for the non-existence of entries is really "bad".  If anything it helps to codify the assumptions, and ensure that our underlying storage is working correctly instead of just saying that it's working.

File: crates/legacy/src/service.rs (hunk 2 above)
Comment (Ayiga) [reply crates/legacy/src/service.rs:4744]: Thanks!

File: crates/legacy/src/service.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (shenkeyao) [reply crates/legacy/src/service.rs:2659]: > We _could_ potentially replace this with a check for the `i = 10`th entry only to achieve the same thing.

Yeah exactly, IIUC I think this is done via the `assert` at line 2676, which is why I think the loop here is a bit redundant. Nothing incorrect though.

[the remark this answers is not in the document]
Comment (Ayiga) [reply crates/legacy/src/service.rs]: Addressed in 81e5660b8d7cf854a774dbc3514af3be67bc4fbe.

**Review by shenkeyao — APPROVED** [review]
